omp/components/prefix_sum.cpp

```diff
@@ -33,21 +33,72 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "core/components/prefix_sum.hpp"
 
 
+#include <algorithm>
+
+
+#include <omp.h>
+
+
+#include "core/base/allocator.hpp"
+
+
 namespace gko {
 namespace kernels {
 namespace omp {
 namespace components {
 
 
+/*
+ * The last entry of the input array is never used, but is replaced.
+ */
 template <typename IndexType>
-void prefix_sum(std::shared_ptr<const OmpExecutor> exec, IndexType *counts,
-                size_type num_entries)
+void prefix_sum(std::shared_ptr<const OmpExecutor> exec,
+                IndexType *const counts, const size_type num_entries)
 {
-    IndexType partial_sum{};
-    for (size_type i = 0; i < num_entries; ++i) {
-        auto nnz = counts[i];
-        counts[i] = partial_sum;
-        partial_sum += nnz;
+    // the operation only makes sense for arrays of size at least 2
+    if (num_entries < 2) {
+        if (num_entries == 0) {
+            return;
+        } else {
+            counts[0] = 0;
+            return;
+        }
+    }
```

Comment:
For 1 entry shouldn't you still return `counts[0]`. In effect, I don't understand the sequential implementation of our prefix sum. It always puts `0` in `counts[0]` and has a lagging behavior. What gives? Usually you would simply do:
```
for (size_type i = 1; i < num_entries ; i++) {
    counts[i] += counts[i-1]; 
}
```

Reply:
We're not actually returning a value, but I guess you're talking about setting `counts[0]`. Since `num_entries` is assumed to be input array size that we are supposed to operate on, it being zero is taken to mean 'do nothing'. If there is one entry, I indeed just do what is equivalent to the reference implementation. IMO, this operation only makes sense for `num_entries >= 2`.

In general, the idea of the implementation is to not only compute a running sum, but also shift the whole array to the right by one position. I thought of exactly the same thing as your loop. But the input/output semantics of wherever this function is used, implies we also need to shift the array and set the first position zero. So, an array containing the lengths of each row can be converted to `row_ptrs` as long as it has an extra allocation in the end. I guess this is referred to as an exclusive scan, as opposed to an inclusive scan.

Reply:
You're not using the `return` statement but `counts` is where what you computed is passed to the user (returned).

Also, if the purpose of this is to be implemented as it is, then I would really not call this `prefix_sum` because the shifting to the right part is not natural and not how actual prefix sums are implemented/used, it's really confusing.
In addition, what about the final value stored in `counts[num_entries-1]`? This one should also be off by `counts[num_entries-1]` won't it, since the sum has this lagging behavior. You would need to store into `counts[num_entries]` to have the actual full sum.

Reply:
I'm not sure about the name `prefix_sum` either. I would prefer `exclusive_scan`; see https://en.cppreference.com/w/cpp/algorithm/exclusive_scan. Indeed, the typical exclusive scan does not take into account the final value, which is simply lost.

How it's normally used in Ginkgo is: an extra slot is left empty at the end of the input array. So for an array of row sizes for `N` rows, one allocates an array of size `N+1` with the first `N` entries containing useful data. Then the `counts[N]` on output will contain the total number of nonzeros in all rows, as we need for `row_ptrs`. It's a consequence of how this function is used currently throughout the code.

Reply:
What we have here is an *inclusive* prefix sum, what you are describing is an *exclusive* prefix sum. Looking at Ginkgo, I don't think we use exclusive prefix sums anywhere, so this choice seems sensible to me. Most calls that need the full sum just use num_elements + 1 as the input size.

Reply:
I mean, I understand that it's useful for building CSR row pointers in particular, but the way it is this is really unlike a  `prefix_sum`, you need to have extra space available and add a `+1` to the element counts, knowing that the last entry will be discarded and everything is shifted... I also did not even see any disclaimer or note anywhere hinting in this direction, either.

Reply:
I did add a comment on top of my function body. But yes, I'd agree with documenting this in `core/components/prefix_sum.hpp`, if that would be better. I'd also be okay with renaming this everywhere to `exclusive_scan`.

Reply:
I guess using `exclusive_scan` or `exclusive_prefix_sum` as a name would solve the issue for me, but when reviewing this I was expected to find the canonical (inclusive) prefix sum which made this really confusing. Another way would be to implement the usual prefix sum and when using it start at index 1 of the array, no?

Reply:
I find the naming *scan* much less intuitive than *prefix sum*, since it neither describes the operation (summing up) nor over which elements it happens (the previous elements, be it inclusive or exclusive). The common use cases for prefix sums, work and data distribution, all lead to simpler implementations when you use an exclusive scan, because you can get both the beginning and end of a range from the same array without special case treatment. Do you see any use cases for inclusive prefix sums in Ginkgo?

Reply:
I'm not saying we need to implement the inclusive one necessarily, we don't seem to require it right now. It's just that for me since my early algorithmic courses, and as far as I can see in many places on the web (starting with wikipedia), `prefix_sum` is reserved for the inclusive form, and does not have data loss and other such pitfalls which you need to take into account for this one. I think at least having a name which reflects that would be helpful for the future, that's all.

Reply:
I actually thought that a prefix sum is the sum of all "prefix" elements, meaning all elements before `i`. So I thought it is always an **"exclusive"** prefix sum.
Maybe a template parameter `bool inclusive` is all we need to have both implementations available? We already have this in [`common/components/prefix_sum.hpp.inc` L44](https://github.com/ginkgo-project/ginkgo/blob/develop/common/components/prefix_sum.hpp.inc#L44)
Also, I agree that we should have the behavior documented in [the prefix header](https://github.com/ginkgo-project/ginkgo/blob/develop/core/components/prefix_sum.hpp)

Reply:
I agree that `scan` is a weird name for this operation. But the advantage of `exclusive_scan` is it's shorter than `exclusive_prefix_sum`, and consistent with std which means it's kind of standard nomenclature - people would either know it or it would help them to look it up. Of course we could also document what it does. I guess we would also change the name of the relevant files if we change this name.

I'm also okay with just leaving the name as it is and adding proper documentation.

Reply:
I believe `scan` is called as such because it's not only a `prefix_sum` but it can also use `max` instead or any other operator. Maybe we could consider going that way as well if we want generality, although I'm not sure we need it now.

Reply:
Oh yes, I see. The C++ std implementation takes an optional BinaryOp. We don't need it right now, but it would be an option to template prefix_sum on a BinaryOp (though it would be a bit complicated). Maybe for now we keep it as it is, and decide on a name and/or add documentation, unless people feel the added generality is worth the extra time and changes.

Reply:
I agree with @Slaedr : I would keep it as it is while documenting the behavior (would be nice if you could do that in this PR).
We can't be as general as `std::scan`with `BinaryOp` because we need to make sure it works for both CPU and GPU.